File: JavaScript/src/Internals/Functions/MathSum/Function_AVERAGEIF.js
import { Function_3 } from '../Function_3.js';
import { Operand } from '../../../Operand.js';
import { FunctionUtil } from '../FunctionUtil.js';

class Function_AVERAGEIF extends Function_3 {
    get Name() {
        return "AverageIf";
    }

    constructor(z) {
        super(z);
    }

    evaluate(engine, tempParameter) {
        let args1 = this.getArray_1(engine, tempParameter); if (args1.IsError) { return args1; }
        let args2 = this.b.evaluate(engine, tempParameter); if (args2.IsError) { return args2; }

        let list = [];
        let o = FunctionUtil.F_base_GetList(args1, list);
        if (o == false) { return this.parameterError(1); }

        let sumdbs;
        if (this.c != null) {
            let args3 = this.getArray_3(engine, tempParameter); if (args3.IsError) { return args3; }
            sumdbs = [];
            let o2 = FunctionUtil.F_base_GetList(args3, sumdbs);
            if (o2 == false) { return this.parameterError(3); }
        } else {
            sumdbs = list;
        }

        let sum;
        let count;
        if (args2.IsNumber) {
            count = FunctionUtil.F_base_countif(list, args2.NumberValue);
            sum = count * args2.NumberValue;
        } else {
            if (args2.IsText) {
                let TextValue = args2.TextValue.trim();
                let parsedValue = parseFloat(TextValue);
                if (!isNaN(parsedValue)) {
                    count = FunctionUtil.F_base_countif(list, parsedValue);
                    sum = FunctionUtil.F_base_sumif(list, parsedValue, sumdbs);
                } else {
                    let sunif = TextValue;
                    let m2 = FunctionUtil.sumifMatch(sunif);
                    if (m2 != null) {
                        count = FunctionUtil.F_base_countif(list, m2.operator, m2.value);
                        sum = FunctionUtil.F_base_sumif(list, m2.operator, m2.value, sumdbs);
                    } else {
                        return this.parameterError(2);
                    }
                }
            } else {
                return this.parameterError(2);
            }
        }
        if (count == 0) {
            return this.div0Error();
        }
        return Operand.Create(sum / count);
    }
}

export { Function_AVERAGEIF };

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/functions/mathSum/Function_AVERAGEIF.java
package toolgood.algorithm.internals.functions.mathSum;

import java.util.AbstractMap.SimpleEntry;
import java.util.List;
import java.util.ArrayList;
import java.math.BigDecimal;
import java.util.function.BiFunction;
import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.NoneEngine;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_3;
import toolgood.algorithm.internals.functions.FunctionUtil;

final class Function_AVERAGEIF extends Function_3 {

    Function_AVERAGEIF(FunctionBase[] funcs) {
        super(funcs);
        if (funcs.length < 2 || funcs.length > 3) {
            throw new IllegalArgumentException("Function '" + Name() + "' requires 2 to 3 parameters.");
        }
    }

    @Override
    public String Name() {
        return "AverageIf";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = GetArray_1(engine, tempParameter); if (args1.IsErrorOrNone()) { return args1; }
        Operand args2 = func2.Evaluate(engine, tempParameter); if (args2.IsErrorOrNone()) { return args2; }

        List<BigDecimal> list = new ArrayList<>();
        boolean o = FunctionUtil.FlattenToList(args1, list);
        if (o == false) { return ParameterError(1); }

        List<BigDecimal> sumdbs;
        if (func3 != null) {
            Operand args3 = GetArray_3(engine, tempParameter); if (args3.IsErrorOrNone()) { return args3; }
            sumdbs = new ArrayList<>();
            boolean o2 = FunctionUtil.FlattenToList(args3, sumdbs);
            if (o2 == false) { return ParameterError(3); }
        } else {
            sumdbs = list;
        }

        BigDecimal sum;
        int count;
        if (args2.IsNumber()) {
            count = FunctionUtil.GetCountIf(list, args2.NumberValue());
            sum = FunctionUtil.GetSumIf(list, args2.NumberValue(), sumdbs);
        } else {
            BigDecimal d = TryParseDecimal(args2.TextValue().trim());
            if (d != null) {
                count = FunctionUtil.GetCountIf(list, d);
                sum = FunctionUtil.GetSumIf(list, d, sumdbs);
            } else {
                SimpleEntry<String, BigDecimal> m2 = FunctionUtil.ParseSumIfMatch(args2.TextValue().trim());
                if (m2 != null) {
                    count = FunctionUtil.GetCountIf(list, m2.getKey(), m2.getValue());
                    sum = FunctionUtil.GetSumIf(list, m2.getKey(), m2.getValue(), sumdbs);
                } else {
                    return ParameterError(2);
                }
            }
        }
        if (count == 0) {
            return Div0Error();
        }
        return Operand.Create(sum.divide(BigDecimal.valueOf(count), java.math.RoundingMode.HALF_UP));
    }

    private static BigDecimal TryParseDecimal(String s) {
        try {
            return new BigDecimal(s);
        } catch (Exception e) {
            return null;
        }
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.ARRAY);
        func2.GetParameterTypes(noneEngine, result, OperandType.NONE);
        if (func3 != null) {
            func3.GetParameterTypes(noneEngine, result, OperandType.ARRAY);
        }
    }
}
